README.md
# comodojo/rpcserver

[![Build Status](https://api.travis-ci.org/comodojo/rpcserver.png)](http://travis-ci.org/comodojo/rpcserver) [![Latest Stable Version](https://poser.pugx.org/comodojo/rpcserver/v/stable)](https://packagist.org/packages/comodojo/rpcserver) [![Total Downloads](https://poser.pugx.org/comodojo/rpcserver/downloads)](https://packagist.org/packages/comodojo/rpcserver) [![Latest Unstable Version](https://poser.pugx.org/comodojo/rpcserver/v/unstable)](https://packagist.org/packages/comodojo/rpcserver) [![License](https://poser.pugx.org/comodojo/rpcserver/license)](https://packagist.org/packages/comodojo/rpcserver) [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/comodojo/rpcserver/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/comodojo/rpcserver/?branch=master) [![Code Coverage](https://scrutinizer-ci.com/g/comodojo/rpcserver/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/comodojo/rpcserver/?branch=master)

This library provides a framework and transport independent implementation of XML and JSON(2.0) RPC server.

It is designed to work in combination with a REST framework that could handle the transport side (such as [Comodojo dispatcher](https://github.com/comodojo/dispatcher.framework)).

Main features are:
- full XMLRPC and JSONRPC(2.0) protocols support, including multicall and batch requests
- embedded introspection methods
- [PSR-3](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md) compliant logging
- payload decoding/econding and encryption
- support multiple signatures per method

Following capabilities are supported out of the box:

- [xmlrpc](http://www.xmlrpc.com/spec)
- [system.multicall](http://www.xmlrpc.com/discuss/msgReader$1208)
- [introspection](http://phpxmlrpc.sourceforge.net/doc-2/ch10.html)
- [nil](http://www.ontosys.com/xml-rpc/extensions.php)
- [faults_interop](http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php)
- [json-rpc](http://www.jsonrpc.org/specification)

Additional capabilities could be implemented [extending the library](#extending-the-library).

## Installation

Install [composer](https://getcomposer.org/), then:

`` composer require comodojo/rpcserver ``

## Basic usage

Quick and dirty example, without a framework that mediates RPC requests:

```php

// get the raw request payload (POST)
$payload = file_get_contents('php://input');

try {

	// create a RpcServer instance (i.e. JSON)
    $server = new \Comodojo\RpcServer\RpcServer(\Comodojo\RpcServer\RpcServer::JSONRPC);
    
    // create a method (using lambda functions)
    $method = \Comodojo\RpcServer\RpcMethod::create("example.sum", function($params) {
    
        $a = $params->get('a');

        $b = $params->get('b');

        return intval($a) + intval($b);

    })->setDescription("Sum two integers")
    ->setReturnType('int')
    ->addParameter('int','a')
    ->addParameter('int','b');
    
    // register method into server
    $is_method_in = $server->methods()->add($method);
    
    // set the payload
    $server->setPayload($request);
    
    // serve the request
    $result = $server->serve();

} catch (\Exception $e) {
	
	/* something did not work :( */

}

echo $result;

```

### Creating methods

The `\Comodojo\RpcServer\RpcMethod` class should be used to create RPC methods to inject into server.

It requires basically a method name and a callable, provided as labmda function, named function or couple class::method.

Parameters can be added using `addParameter()` method. Multiple signatures could be specified using `addSignature()`.

For example, to create a **my.method** RPC method mapped to `\My\RpcClass::mymethod()` that has two different signatures:

```php

// create a method using class::method pattern
$method = \Comodojo\RpcServer\RpcMethod::create("my.method", "\My\RpcClass", "mymethod")
    // provide a description for the method
    ->setDescription("My method")
    // set the return type (default: undefined)
    ->setReturnType('boolean')
    // add a second signature
    ->addSignature()
    // set the return type for second signature
    ->setReturnType('boolean')
    // add expected parameters (if any) for second signature
    ->addParameter('int','a')
    ->addParameter('int','b');

```

Signatures are automatically matched by server as well as received parameters; if a request does not match any valid signature, a "Invalid params" (-32602) error is returned back to client.

The `\My\RpcClass::mymethod()` should expect a `\Comodojo\RpcServer\Request\Parameters` object that provides received parameters, server properties and logging interface.

```php

class RpcClass {

    public function mymethod($params) {
    
        $a = $params->get('a');
        
        $b = $params->get('b');
        
        $logger = $params->logger();
        
        $current_rpc_protocol = $params->protocol();
        
        $return = array($a, $b);
        
        $logger->info("Current protocol: ".$current_rpc_protocol.", returned value in context", $return);
        
        return $return;
    
    }

}

```

### Server properties

The `\Comodojo\RpcServer\RpcServer` should be created specifying the RPC protocol (using constant `RpcServer::JSONRPC` or `RpcServer::XMLRPC`).

The optional parameter *$logger* expects an implementation of `\Psr\Log\LoggerInterface` and enable internal logging.

```php

// create the server
$server = new RpcServer(RpcServer::XMLRPC);

// (optional) set encoding (default to *utf-8*)
$server->setEncoding('ISO-8859-1');

```

Once created, server expects a payload, and starts processing it when `serve()` is invoked:

```php

$result = $server->setPayload($request)->serve();

```

A non-standard PSK encryption using AES could be enabled specifying key with `$server->setEncryption($key)`; the [Comodojo RpcClient](https://github.com/comodojo/rpcclient) package implements a compatible client.

## Extending the library

RPC methods, server capabilities and standard errors can be added/removed using:

- `$server->methods()`
- `$server->capabilities()`
- `$server->errors()`

For example, to add a server capability:

```php

$my_capability = $server->capabilities()->add("my.capability", "http://url.to.my/capability", 1.0);

```

## Documentation

- [API](https://api.comodojo.org/libs/Comodojo/RpcServer.html)

## Contributing

Contributions are welcome and will be fully credited. Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## License

`` comodojo/rpcserver `` is released under the MIT License (MIT). Please see [License File](LICENSE) for more information.

CHANGELOG.md
# Changelog

## Version 1.0.0

- Initial stable release

## Version 1.0.0-beta1

- first beta pre-release

src/RpcServer.php
<?php namespace Comodojo\RpcServer;

use \Comodojo\RpcServer\Component\Capabilities;
use \Comodojo\RpcServer\Component\Methods;
use \Comodojo\RpcServer\Component\Errors;
use \Comodojo\RpcServer\Request\Parameters;
use \Comodojo\RpcServer\Request\XmlProcessor;
use \Comodojo\RpcServer\Request\JsonProcessor;
use \Comodojo\RpcServer\Util\NullLogger;
use \Comodojo\Xmlrpc\XmlrpcEncoder;
use \Comodojo\Xmlrpc\XmlrpcDecoder;
use \Crypt_AES;
use \Comodojo\Exception\RpcException;
use \Comodojo\Exception\XmlrpcException;
use \Exception;


/** 
 * The RpcServer main class.
 * 
 * @package     Comodojo Spare Parts
 * @author      Marco Giovinazzi <marco.giovinazzi@comodojo.org>
 * @license     MIT
 *
 * LICENSE:
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
 
class RpcServer {

    /**
     * Capabilities collector
     *
     * @const string
     */
    const XMLRPC = 'xml';
    
    /**
     * Capabilities collector
     *
     * @const string
     */
    const JSONRPC = 'json';

    /**
     * Capabilities collector
     *
     * @var \Comodojo\RpcServer\Component\Capabilities
     */
    private $capabilities = null;
    
    /**
     * RpcMethods collector
     *
     * @var \Comodojo\RpcServer\Component\Methods
     */
    private $methods = null;
    
    /**
     * Standard Rpc Errors collector
     *
     * @var \Comodojo\RpcServer\Component\Errors
     */
    private $errors = null;
    
    /**
     * The request payload, better the RAW export of 'php://input'
     *
     * @var string
     */
    private $payload = null;
    
    /**
     * Encryption key, in case of encrypted transport
     *
     * @var string
     */
    private $encrypt = null;
    
    /**
     * Current encoding
     *
     * @var string
     */
    private $encoding = 'utf-8';
    
    /**
     * Current protocol
     *
     * @var string
     */
    private $protocol = null;
    
    /**
     * Supported RPC protocols
     *
     * @var array
     */
    private $supported_protocols = array('xml', 'json');

    /**
     * Internal marker (encryption)
     *
     * @var bool
     */
    private $request_is_encrypted = false;

    /**
     * Current logger
     *
     * @var \Psr\Log\LoggerInterface
     */
    private $logger = null;
    
    /**
     * Class constructor
     *
     * @param string $protocol
     *
     * @throws \Exception
     */
    public function __construct($protocol, \Psr\Log\LoggerInterface $logger = null) {
        
        $this->logger = is_null($logger) ? new NullLogger() : $logger;

        try {
            
            // setup protocol

            $this->setProtocol($protocol);

            // init components

            $this->capabilities = new Capabilities($this->logger);
        
            $this->methods = new Methods($this->logger);
            
            $this->errors = new Errors($this->logger);

            // populate components

            self::setIntrospectionMethods($this->methods);
            
            self::setCapabilities($this->capabilities);

            self::setErrors($this->errors);
            
        } catch (Exception $e) {
            
            throw $e;
            
        }
        
        $this->logger->notice("RpcServer init complete, protocol ".$protocol);
        
    }
    
    /**
     * Set RPC protocol (json or xml)
     *
     * @param string $protocol
     *
     * @return \Comodojo\RpcServer\RpcServer
     * @throws \Exception
     */
    final public function setProtocol($protocol) {
        
        if ( empty($protocol) || !in_array($protocol, $this->supported_protocols) ) throw new Exception('Invalid or unsupported RPC protocol');
        
        $this->protocol = $protocol;
        
        return $this;
        
    }
    
    /**
     * Get RPC protocol
     *
     * @return string
     */
    final public function getProtocol() {
        
        return $this->protocol;
        
    }
    
    /**
     * Set request payload, raw format
     *
     * @return \Comodojo\RpcServer\RpcServer
     */
    final public function setPayload($payload) {
        
        $this->payload = $payload;
        
        return $this;
        
    }
    
    /**
     * Get request payload
     *
     * @return string
     */
    final public function getPayload() {
        
        return $this->payload;
        
    }
    
    final public function setEncoding($encoding) {
        
        $this->encoding = $encoding;
        
        return $this;
        
    }
    
    final public function getEncoding() {
        
        return $this->encoding;
        
    }
    
    /**
     * Set encryption key; this will enable the NOT-STANDARD payload encryption
     *
     * @param   string  $key Encryption key
     *
     * @return  RpcServer
     * 
     * @throws \Exception
     */
    final public function setEncryption($key) {

        if ( empty($key) ) throw new Exception("Shared key cannot be empty");

        $this->encrypt = $key;

        return $this;

    }
    
    /**
     * Get the ecryption key or null if no encryption is selected
     *
     * @return string
     */
    final public function getEncryption() {
        
        return $this->encrypt;
        
    }
    
    /**
     * Get the capabilities manager
     *
     * @return \Comodojo\RpcServer\Component\Capabilities
     */
    public function capabilities() {

        return $this->capabilities;

    }
    
    /**
     * Get the methods manager
     *
     * @return \Comodojo\RpcServer\Component\Methods
     */
    public function methods() {

        return $this->methods;

    }
    
    /**
     * Get the errors manager
     *
     * @return \Comodojo\RpcServer\Component\Errors
     */
    public function errors() {

        return $this->errors;
    }

    /**
     * Retrieve the logger instance
     *
     * @return \Psr\Log\LoggerInterface
     */
    public function logger() {

        return $this->logger;

    }
    
    /**
     * Serve request
     *
     * @return string
     * @throws Exception
     */
    public function serve() {

        $this->logger->notice("Start serving request");
        
        $parameters_object = new Parameters($this->capabilities, $this->methods, $this->errors, $this->logger, $this->protocol);
        
        try {

            $this->logger->debug("Received payload: ".$this->payload);
            
            $payload = $this->uncan($this->payload);

            $this->logger->debug("Decoded payload", (array)$payload);

            if ( $this->protocol == self::XMLRPC ) $result = XmlProcessor::process($payload, $parameters_object, $this->logger);
            
            else if ( $this->protocol == self::JSONRPC ) $result = JsonProcessor::process($payload, $parameters_object, $this->logger);
            
            else throw new Exception('Invalid or unsupported RPC protocol');
            
        } catch (RpcException $re) {
            
            return $this->can($re, true);
            
        } catch (Exception $e) {
            
            throw $e;
            
        }
        
        return $this->can($result, false);
        
    }
    
    /**
     * Uncan the provided payload
     *
     * @param string $payload
     * 
     * @return mixed
     * @throws \Comodojo\Exception\RpcException
     */
    private function uncan($payload) {
        
        $decoded = null;
        
        if ( empty($payload) || !is_string($payload) ) throw new RpcException("Invalid Request", -32600);
        
        if ( substr($payload, 0, 27) == 'comodojo_encrypted_request-' ) {
            
            if ( empty($this->encrypt) ) throw new RpcException("Transport error", -32300);
            
            $this->request_is_encrypted = true;
            
            $aes = new Crypt_AES();
            
            $aes->setKey($this->encrypt);
            
            $payload = $aes->decrypt(base64_decode(substr($payload, 27)));
            
            if ( $payload == false ) throw new RpcException("Transport error", -32300);
            
        }
        
        if ( $this->protocol == 'xml' ) {
            
            $decoder = new XmlrpcDecoder();
            
            try {

                $decoded = $decoder->decodeCall($payload);
                
            } catch (XmlrpcException $xe) {
                
                throw new RpcException("Parse error", -32700);
                
            }
            
        } else if ( $this->protocol == 'json' ) {

            if ( strtolower($this->encoding) != 'utf-8' ) {

                $payload = mb_convert_encoding($payload, "UTF-8", strtoupper($this->encoding));

            }
            
            $decoded = json_decode($payload, false /*DO RAW conversion*/);
            
            if ( is_null($decoded) ) throw new RpcException("Parse error", -32700);
            
        } else {
            
            throw new RpcException("Transport error", -32300);
            
        }
        
        return $decoded;
        
    }
    
    /**
     * Can the RPC response
     *
     * @param mixed   $response
     * @param boolean $error
     * 
     * @return string
     * @throws \Comodojo\Exception\RpcException
     */
    private function can($response, $error) {
        
        $encoded = null;
        
        if ( $this->protocol == 'xml' ) {
            
            $encoder = new XmlrpcEncoder();

            $encoder->setEncoding($this->encoding);
            
            try {

                $encoded = $error ? $encoder->encodeError($response->getCode(), $response->getMessage()) : $encoder->encodeResponse($response);
                
            } catch (XmlrpcException $xe) {
                
                $encoded = $encoder->encodeError(-32500, "Application error");

            }
            
        } else {

            if ( strtolower($this->encoding) != 'utf-8' && !is_null($response) ) {

                array_walk_recursive($response, function(&$entry) {
                    
                    if ( is_string($entry) ) {

                        $entry = mb_convert_encoding($entry, strtoupper($this->encoding), "UTF-8");

                    }

                });
                
            }
            
            // json will not return any RpcException; errors (in case) are handled directly by processor
            
            $encoded = is_null($response) ? null : json_encode($response/*, JSON_NUMERIC_CHECK*/);
            
        }
        
        if ( $this->request_is_encrypted /* && !empty($encoded) */ ) {
            
            $aes = new Crypt_AES();
            
            $aes->setKey($this->encrypt);
            
            $encoded = 'comodojo_encrypted_response-'.base64_encode($aes->encrypt($encoded));
            
        }
        
        return $encoded;
        
    }
    
    /**
     * Inject introspection and reserved RPC methods
     *
     * @param \Comodojo\RpcServer\Component\Methods $methods
     */
    private static function setIntrospectionMethods($methods) {
        
        $methods->add( RpcMethod::create("system.getCapabilities", "\Comodojo\RpcServer\Reserved\GetCapabilities", "execute")
            ->setDescription("This method lists all the capabilites that the RPC server has: the (more or less standard) extensions to the RPC spec that it adheres to")
            ->setReturnType('struct')
        );
        
        $methods->add( RpcMethod::create("system.listMethods", "\Comodojo\RpcServer\Introspection\ListMethods", "execute")
            ->setDescription("This method lists all the methods that the RPC server knows how to dispatch")
            ->setReturnType('array')
        );
            
        $methods->add( RpcMethod::create("system.methodHelp", "\Comodojo\RpcServer\Introspection\MethodHelp", "execute")
            ->setDescription("Returns help text if defined for the method passed, otherwise returns an empty string")
            ->setReturnType('string')
            ->addParameter('string', 'method')
        );
            
        $methods->add( RpcMethod::create("system.methodSignature", "\Comodojo\RpcServer\Introspection\MethodSignature", "execute")
            ->setDescription("Returns an array of known signatures (an array of arrays) for the method name passed.". 
                "If no signatures are known, returns a none-array (test for type != array to detect missing signature)")
            ->setReturnType('array')
            ->addParameter('string', 'method')
        );
        
        $methods->add( RpcMethod::create("system.multicall", "\Comodojo\RpcServer\Reserved\Multicall", "execute")
            ->setDescription("Boxcar multiple RPC calls in one request. See http://www.xmlrpc.com/discuss/msgReader\$1208 for details")
            ->setReturnType('array')
            ->addParameter('array', 'requests')
        );

    }
    
    /**
     * Inject supported capabilities
     *
     * @param \Comodojo\RpcServer\Component\Capabilities $capabilities
     */
    private static function setCapabilities($capabilities) {

        $supported_capabilities = array(
            'xmlrpc' => array('http://www.xmlrpc.com/spec', 1),
            'system.multicall' => array('http://www.xmlrpc.com/discuss/msgReader$1208', 1),
            'introspection' => array('http://phpxmlrpc.sourceforge.net/doc-2/ch10.html', 2),
            'nil' => array('http://www.ontosys.com/xml-rpc/extensions.php', 1),
            'faults_interop' => array('http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php', 20010516),
            'json-rpc' => array('http://www.jsonrpc.org/specification', 2)
        );
    
        foreach ($supported_capabilities as $capability => $values) {
            
            $capabilities->add($capability, $values[0], $values[1]);

        }
    
    }

    /**
     * Inject standard and RPC errors
     *
     * @param \Comodojo\RpcServer\Component\Errors $errors
     */
    private static function setErrors($errors) {

        $std_rpc_errors = array(
            -32700 => "Parse error",
            -32701 => "Parse error - Unsupported encoding",
            -32702 => "Parse error - Invalid character for encoding",
            -32600 => "Invalid Request",
            -32601 => "Method not found",
            -32602 => "Invalid params",
            -32603 => "Internal error",
            -32500 => "Application error",
            -32400 => "System error",
            -32300 => "Transport error",
            // Predefined Comodojo Errors
            -31000 => "Multicall is available only in XMLRPC",
            -31001 => "Recursive system.multicall forbidden"
        );

        foreach ($std_rpc_errors as $code => $message) {
            
            $errors->add($code, $message);

        }
        
    }
    
 }
 